src/io/bin.cpp

```diff
@@ -150,8 +150,10 @@ namespace LightGBM {
   }
 
   std::vector<double> FindBinWithZeroAsOneBin(const double* distinct_values, const int* counts,
-    int num_distinct_values, int max_bin, size_t total_sample_cnt, int min_data_in_bin) {
+    int num_distinct_values, int max_bin, size_t total_sample_cnt, int min_data_in_bin, std::vector<double> forced_upper_bounds) {
```

Comment:
could we add an independent function,`FindBinWithPredefinedBin`, to include these changes? I am afraid of these changes may introduce some bugs.

Reply:
@guolinke Yes, I can make that change.

Reply:
ping @btrotta

Reply:
@guolinke I think that would result in sub-optimal choice of bins. Currently my code takes into account the forced bins when choosing the remaining ones, so that we can find evenly sized bins (i.e. it will not choose thresholds too close to the forced thresholds). But if we use `FindBinWithZeroAsOneBin` to create the non-forced thresholds, the thresholds it chooses may be very close to the forced thresholds in which case the bins will not be evenly sized.